packages/date/src/index.js

```diff
@@ -409,7 +414,17 @@ export function format( dateFormat, dateValue = new Date() ) {
 	}
 	// Join with [] between to separate characters, and replace
 	// unneeded separators with static text.
-	return momentDate.format( newFormat.join( '[]' ) );
+	const dateString = momentDate.format( newFormat.join( '[]' ) );
+	/**
+	 * Filters the formatted date like 'wp_date()'
+	 *
+	 * @param {string} dateString Formatted result.
+	 * @param {string} dateFormat Format string.
+	 * @param {Moment} momentDate Parsed date by moment.js.
+	 */
+	return String(
+		applyFilters( 'date.format', dateString, dateFormat, momentDate )
```

Comment:
Rather than filtering the format late, did you look into changing the dateFormat value that is passed to this function? It might be cleaner if the change could be applied as a setting directly. In the past the team has tried to avoid hooks in favor of a more declarative/reactive API. Maybe @gziolo has a suggestion for a better way to extend this

Reply:
> Rather than filtering the format late, did you look into changing the `dateFormat` value that is passed to this function?

Yeah, I'm not sure if the string after being passed to this function needs to be formatted (eg. `F j, KX` -> `January 26, KX` -> I need new format string `K` and `X` but... reformat?)

> In the past the team has tried to avoid hooks in favor of a more declarative/reactive API. Maybe @gziolo has a suggestion for a better way to extend this

I got it. Thank you for your review.

Reply:
> Rather than filtering the format late, did you look into changing the dateFormat value that is passed to this function? It might be cleaner if the change could be applied as a setting directly.

That would be ideal, but you also can't control all places where it is used 👍🏻 

We have introduced filters for i18n functions as explained in:

https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/filters/i18n-filters.md

Implementation can be found in:

https://github.com/WordPress/gutenberg/blob/trunk/packages/i18n/src/create-i18n.js

If we were to introduce any filters, we should figure out a good strategy to cover all methods where it makes sense.

For dates, you can also customize all settings with `setSettings` call so maybe there aren't that many places where those filters make sense.